Add support for readonly class in PHP 8.2

ngmy opened the pull request:
#235 のPRです。

## Summary by Sourcery

Add support for readonly classes in PHP 8.2.

New Features:
- Support readonly classes in PHP 8.2

Tests:
- Added a test for readonly class instantiation

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Enhanced aspect-oriented capabilities that automatically select the optimal interception behavior based on the PHP version, fully supporting PHP 8.2’s readonly properties.
	- Improved state management for more consistent and flexible behavior across different environments.
	- Introduced a new readonly class to demonstrate PHP 8.2 compatibility.
	- Added a new method to the `WeavedInterface` for improved binding initialization.

- **Tests**
	- Added comprehensive tests to verify PHP 8.2 compatibility, including scenarios for readonly classes and enhanced interception functionality.
	- Updated tests to reflect changes in state management and bindings access.
	- Introduced new test cases to validate the instantiation of classes compatible with PHP 8.2.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ngmy committed:
Add support for readonly class in PHP 8.2

koriym committed:
Update PHPDoc annotations to use imported MethodBindings type

Replaced verbose array annotations with the `MethodBindings` type imported via Psalm across interfaces and traits. This improves code readability and ensures type consistency throughout the project.

koriym committed:
Rename `initState` method for not conflict with the method names of the AOP class that inherits it

koriym commented on src/WeavedInterface.php:
```diff
@@ -4,6 +4,13 @@
+     *
+     * @SuppressWarnings(PHPMD.CamelCaseMethodName)
+     */
+    public function _initState(array $bindings): void;
```
The proposal seems strange. See @https://phpmd.org/documentation/suppress-warnings.html

koriym commented on src/InterceptTrait.php:
```diff
@@ -8,19 +8,26 @@
+    public function _initState(array $bindings): void
+    {
+        $this->state = new InterceptTraitState($bindings, true);
+    }
```
This is intentional to prevent inherited classes from conflicting.

koriym commented on src-php8/Php82InterceptTrait.php:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @SuppressWarnings(PHPMD.CamelCaseMethodName)
+     */
+    private function _intercept(string $func, array $args) // phpcs:ignore
+    {
+        if (! $this->state->isAspect) {
+            $this->state->isAspect = true;
+
+            return call_user_func_array([parent::class, $func], $args);
+        }
+
+        $this->state->isAspect = false;
+        $result = (new Invocation($this, $func, $args, $this->state->bindings[$func]))->proceed();
+        $this->state->isAspect = true;
+
+        return $result;
+    }
```
`_intercept` is a special method for AOP, so there is no need to check its robustness. Simplicity and speed are prioritized.

koriym commented on src/InterceptTrait.php:
```diff
@@ -31,15 +36,15 @@ trait InterceptTrait
+        $this->state->isAspect = true;
 
         return $result;
     }
```
`_intercept` is a special method for AOP, so there is no need to check its robustness. Simplicity and speed are prioritized.

koriym commented on src/InterceptTraitState.php:
```diff
@@ -0,0 +1,27 @@
+
+final class InterceptTraitState
+{
+    /**
+     * @var array<string, array<class-string<MethodInterceptor>>>
+     * @readonly
+     */
+    public $bindings;
```
This package supports PHP 7.2 or higher.

koriym commented:
@sourcery-ai guide in Japanese

koriym commented:
@ngmy 
- メソッド名が被らないようにメソッド名を`_initState`にしました。
- psalmのドメインタイプ `MethodBindings`を適用しました。

koriym committed:
Add red test

Add $state property to FakeMockChild class

koriym committed:
Rename `state` to `_state` to avoid conflicts

This change updates the property name from `state` to `_state` across relevant classes and tests to reinforce encapsulation and align with internal naming conventions. Adjustments were made to all associated references to ensure consistency and functionality.

koriym committed:
Refactor `_state` property to private visibility.

Changed `_state` visibility from public to private in `InterceptTrait` and `Php82InterceptTrait`. Updated tests to remove direct modifications to `_state`, ensuring encapsulation and adherence to best practices.

koriym commented on src/InterceptTrait.php:
```diff
@@ -11,16 +11,21 @@
+     * @internal Public for CompilerTest
      */
-    public $bindings = [];
+    public $state;
```
Make it private.

koriym replied:
Fixed.

koriym commented on src/InterceptTrait.php:
```diff
@@ -8,19 +8,25 @@
      * @readonly
      */
-    public $bindings = [];
+    private $_state;
```
> @readonly annotation suggests PHP 8.2 features
No.

koriym committed:
Simplify constructor for InterceptTraitState

Removed the unnecessary `isAspect` parameter from the InterceptTraitState constructor and set its default value directly in the class property. This simplifies the initialization process and reduces redundant arguments.

koriym committed:
Refactor type annotations in Php82InterceptTrait

Replaced explicit type definitions with imported Psalm type aliases for better readability and maintainability. Updated method parameter annotations to use `MethodBindings` and `Arguments` from the `Types` class.

koriym commented:
@ngmy これでどうでしょうか。変更分のreviewをお願いします！

ngmy commented:
@koriym LGTMです！
